fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipIndexManager.java

```diff
@@ -0,0 +1,602 @@
+            final boolean deleteProperties) {
+        // End all membership added in this transaction
+        final Map<String, Object> parameterSource = Map.of(
+                "txId", txId,
+                "sourceId", sourceId.getFullId(),
+                "addOp", ADD_OPERATION);
+
+        jdbcTemplate.update(CLEAR_ALL_ADDED_FOR_SOURCE_IN_TX, parameterSource);
+
+        // End all membership that existed prior to this transaction
+        if (deleteProperties) {
+            final Map<String, Object> parameterSource2 = Map.of(
+                    "txId", txId,
+                    "sourceId", sourceId.getFullId(),
+                    "forceFlag", FORCE_FLAG,
+                    "deleteOp", DELETE_OPERATION);
+            jdbcTemplate.update(DELETE_EXISTING_FOR_SOURCE, parameterSource2);
+        } else {
+            final Map<String, Object> parameterSource2 = Map.of(
+                    "txId", txId,
+                    "sourceId", sourceId.getFullId(),
+                    "endTime", Timestamp.from(endTime),
+                    "noEndTime", NO_END_TIMESTAMP,
+                    "deleteOp", DELETE_OPERATION);
+            jdbcTemplate.update(END_EXISTING_FOR_SOURCE, parameterSource2);
+        }
+    }
+
+    /**
+     * Clean up any references to the target id, in transactions and outside
+     * @param txId transaction id
+     * @param targetId identifier of the resource to cleanup membership references for
+     */
+    public void deleteMembershipReferences(final String txId, final FedoraId targetId) {
+        final Map<String, Object> parameterSource = Map.of(
+                "targetId", targetId.getFullId(),
+                "txId", txId);
+
+        jdbcTemplate.update(PURGE_ALL_REFERENCES_TRANSACTION, parameterSource);
+        jdbcTemplate.update(PURGE_ALL_REFERENCES_MEMBERSHIP, parameterSource);
+    }
+
+    /**
+     * Add new membership property to the index, clearing any delete
+     * operations for the property if necessary.
+     * @param txId transaction id
+     * @param sourceId ID of the direct/indirect container which produced the membership
+     * @param membership membership triple
+     * @param startTime time the membership triple was added
+     */
+    public void addMembership(final String txId, final FedoraId sourceId, final Triple membership,
+            final Instant startTime) {
+        // Clear any existing delete operation for this membership
+        final Map<String, Object> parametersDelete = Map.of(
+                "txId", txId,
+                "sourceId", sourceId.getFullId(),
+                "subjectId", membership.getSubject().getURI(),
+                "property", membership.getPredicate().getURI(),
+                "objectId", membership.getObject().getURI(),
+                "operation", DELETE_OPERATION);
+
+        jdbcTemplate.update(CLEAR_ENTRY_IN_TX, parametersDelete);
```

Comment:
I think there may be some oddness here when `addMembership()` is called after calling `deleteMembershipForSource()`, but it's possible that I just don't understand the intended behavior. The problem I see is that there are two different types of deletes, with or without the `force` flag. `deleteMembershipForSource()` creates deletes with the `force` flag, and I assume that you would always want to apply them. However, this method will delete all of these operations in the transaction, so they won't be applied. Won't this result in duplicate entries (assuming there aren't any db constraints being violated)? I think the case where this matters is MembershipServiceImpl#L104.

Reply:
That's a good point, I believe I should exclude `force_flag` deletes from being cleaned up here.